### ISSUE TITLE
Metrics Performance is Trivial to Fix! (Analysis of the 5.3 -> 5.4 Regression Inside)

### ISSUE THREAD

**Issue by original-brownbear** [elastic/logstash#7904]
Long story short: **The issue with the metrics performance regression is trivial to fix.**

The mistake made in going from `5.3` to `5.4` was that things moved from:

* Lookup and cache counter from `MetricStore` and call e.g. `counter.increment(5)` on it

to 

* Lookup the counter for every change of the metric + use dynamic `send` to invoke the metric method

So the way to fix the metrics performance is simply to:

Go to all places where we use `NameSpacedMetric` and replace the instantiation of it by getting the correct counter (`Gauge`-type object) from `MetricStore` using `fetch_or_store` and cache it.

Also, it's probably a good idea to replace `Gauge.execute` which is currently this:

```rb
    def execute(action, value = nil)
      send(action, value)
    end
```

with actual methods that pass values to the Java logic, `send` will never be optimized here by the JIT I think.

Lots of typing but pretty easy. Note that I did this test wise just for the output counter in the wrapped batch writer and saw about a 20% increase in throughput ... this is def. where the problems are coming from.

Obviously, part of it is also that we are running more metrics now, but if we just cache all counters we should be transparent no matter what. The cost for actually incrementing counters shows as `~0%` in profiling :)

**Comment by original-brownbear** [comment]
@suyograo FYI, you were right all along there was a huge regression in metrics :( Trivial but a little time consuming to fix (pretty safe fix though, could probably back port it all the way).

**Comment by jakelandis** [comment]
@original-brownbear - I am pretty comfortable with this code, there are are couple of key places that I can quickly (and confidently) change from `send` to the actual method.

**Comment by original-brownbear** [comment]
@jakelandis note that the `send` is a much smaller part of the overhead compared to the lookups. If I comment out the send call I get a visible maybe ~15% increase in throughput, if I comment out the lookups get a 3x speedup in raw pipeline throughput.
The lookups are def key here and they weren't even perfectly cached in `5.3`, just a bit less indirection already made a huge impact there.

**Comment by jakelandis** [comment]
@original-brownbear - the `send` is pretty trivial and safe to fix. However caching the metrics outside of the metrics store would require keeping those caches in sync with the pruning (cache invalidation) behavior when restarting a pipeline. The lookup behavior will be removed with #7788, but it may be a bit before that is complete, and is only targeted for 6.x.

**Comment by original-brownbear** [comment]
@jakelandis ah the damned restart ... that's unfortunate :( It's kinda annoying to keep such a massive performance issue around though,  let's see what ohers think later :)

**Comment by original-brownbear** [comment]
Actually, reload is trivial to fix too. Just keep a sequence number in `MetricStore` and each `Metric`, on reload you increment it + check it on every metric operation (since the `Metric` has a reference to the store anyhow) => simply reset count to `0` and increase the sequence number in the metric afterwards on sequence number mismatch and done? Should be like 10 lines of code to get around that?

**Comment by jakelandis** [comment]
Hmm... removing `send` in favor of calling the concrete method doesn't seem to have any measurable impact.  A 2s improvement here may just be statistical noise.

With change:
```
java -cp './build/libs/benchmark-cli.jar:*' org.logstash.benchmark.cli.Main --testcase=baseline --git-hash=jakelandis#43a321093b89f84aa3b4e7bbd98ba81295521343 --repeat-data 10
```
```
Logstash Benchmark
------------------------------------------
Benchmarking Version: jakelandis#43a321093b89f84aa3b4e7bbd98ba81295521343
Running Test Case: baseline (x10)
------------------------------------------
Start Time: Fri 8 4 12:03:36 2017 CDT
Statistical Summary:

Elapsed Time: 175s
Num Events: 9978751
Throughput Min: 0.00
Throughput Max: 332676.00
Throughput Mean: 83156.26
Throughput StdDev: 24571.46
Throughput Variance: 603756766.41
Mean CPU Usage: 24.68%
```

Commit prior to change:
```
java -cp './build/libs/benchmark-cli.jar:*' org.logstash.benchmark.cli.Main --testcase=baseline --git-hash=532a0304e91e5a3d763feff50ab09c56773ef831 --repeat-data 10
```
```
Logstash Benchmark
------------------------------------------
Benchmarking Version: 532a0304e91e5a3d763feff50ab09c56773ef831
Running Test Case: baseline (x10)
------------------------------------------

Start Time: Fri 8 4 12:11:14 2017 CDT
Statistical Summary:

Elapsed Time: 177s
Num Events: 9985111
Throughput Min: 0.00
Throughput Max: 367712.00
Throughput Mean: 81845.17
Throughput StdDev: 27127.60
Throughput Variance: 735906453.45
Mean CPU Usage: 23.85%
```

**Comment by original-brownbear** [comment]
fixed in #7919
